xcm/xcm-builder/tests/scenarios.rs

```diff
@@ -0,0 +1,314 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+mod mock;
+
+use frame_support::weights::Weight;
+use mock::{
+	kusama_like_with_balances, AccountId, Balance, Balances, BaseXcmWeight, XcmConfig, CENTS,
+};
+use polkadot_parachain::primitives::Id as ParaId;
+use sp_runtime::traits::AccountIdConversion;
+use xcm::latest::prelude::*;
+use xcm_executor::XcmExecutor;
+
+pub const ALICE: AccountId = AccountId::new([0u8; 32]);
+pub const PARA_ID: u32 = 2000;
+pub const INITIAL_BALANCE: u128 = 100_000_000_000;
+pub const REGISTER_AMOUNT: Balance = 10 * CENTS;
+
+// Construct a `BuyExecution` order.
+fn buy_execution<C>(debt: Weight) -> Order<C> {
+	use xcm::latest::prelude::*;
+	Order::BuyExecution {
+		fees: (Here, REGISTER_AMOUNT).into(),
+		weight: 0,
+		debt,
+		halt_on_error: false,
+		orders: vec![],
+		instructions: vec![],
+	}
+}
+
+/// Scenario:
+/// A parachain transfers funds on the relaychain to another parachain's account.
+///
+/// Asserts that the parachain accounts are updated as expected.
+#[test]
+fn withdraw_and_deposit_works() {
+	let para_acc: AccountId = ParaId::from(PARA_ID).into_account();
+	let balances = vec![(ALICE, INITIAL_BALANCE), (para_acc.clone(), INITIAL_BALANCE)];
+	kusama_like_with_balances(balances).execute_with(|| {
+		let other_para_id = 3000;
+		let amount = REGISTER_AMOUNT;
+		let weight = 3 * BaseXcmWeight::get();
+		let r = XcmExecutor::<XcmConfig>::execute_xcm(
+			Parachain(PARA_ID).into(),
+			Xcm::WithdrawAsset {
+				assets: vec![(Here, amount).into()].into(),
+				effects: vec![
+					buy_execution(weight),
+					Order::DepositAsset {
+						assets: All.into(),
+						max_assets: 1,
+						beneficiary: Parachain(other_para_id).into(),
+					},
+				],
+			},
+			weight,
+		);
+		assert_eq!(r, Outcome::Complete(weight));
+		let other_para_acc: AccountId = ParaId::from(other_para_id).into_account();
+		assert_eq!(Balances::free_balance(para_acc), INITIAL_BALANCE - amount);
+		assert_eq!(Balances::free_balance(other_para_acc), amount);
+	});
+}
+
+/// Scenario:
+/// A parachain wants to be notified that a transfer worked correctly.
+/// It sends a `QueryHolding` after the deposit to get notified on success.
+/// This somewhat abuses `QueryHolding` as an indication of execution success. It works because
+/// order execution halts on error (so no `QueryResponse` will be sent if the previous order failed).
+/// The inner response is not used.
```

Comment:
```suggestion
/// A parachain wants to be notified that a transfer worked correctly.
/// It includes a `QueryHolding` order after the deposit to get notified on success.
/// This somewhat abuses `QueryHolding` as an indication of execution success. It works because
/// order execution halts on error (so no `QueryResponse` will be sent if the previous order failed).
/// The inner response sent due to the query is not used.
```